pkg/ingester/metrics.go

```diff
@@ -28,93 +29,173 @@ type ingesterMetrics struct {
+		}),
+
+		// Chunks flushing.
+		chunkUtilization: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_ingester_chunk_utilization",
+			Help:    "Distribution of stored chunk utilization (when stored).",
+			Buckets: prometheus.LinearBuckets(0, 0.2, 6),
+		}),
+		chunkLength: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_ingester_chunk_length",
+			Help:    "Distribution of stored chunk lengths (when stored).",
+			Buckets: prometheus.ExponentialBuckets(5, 2, 11), // biggest bucket is 5*2^(11-1) = 5120
+		}),
+		chunkSize: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_ingester_chunk_size_bytes",
+			Help:    "Distribution of stored chunk sizes (when stored).",
+			Buckets: prometheus.ExponentialBuckets(500, 2, 5), // biggest bucket is 500*2^(5-1) = 8000
+		}),
+		chunksPerUser: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_chunks_stored_total",
+			Help: "Total stored chunks per user.",
+		}, []string{"user"}),
+		chunkSizePerUser: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_chunk_stored_bytes_total",
+			Help: "Total bytes stored in chunks per user.",
+		}, []string{"user"}),
+		chunkAge: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_chunk_age_seconds",
+			Help: "Distribution of chunk ages (when stored).",
+			// with default settings chunks should flush between 5 min and 12 hours
+			// so buckets at 1min, 5min, 10min, 30min, 1hr, 2hr, 4hr, 10hr, 12hr, 16hr
+			Buckets: []float64{60, 300, 600, 1800, 3600, 7200, 14400, 36000, 43200, 57600},
+		}),
+		memoryChunks: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ingester_memory_chunks",
+			Help: "The total number of chunks in memory.",
+		}),
+		flushReasons: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_flush_reasons",
+			Help: "Total number of series scheduled for flushing, with reasons.",
+		}, []string{"reason"}),
+		droppedChunks: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ingester_dropped_chunks_total",
+			Help: "Total number of chunks dropped from flushing because they have too few samples.",
+		}),
+		oldestUnflushedChunkTimestamp: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_oldest_unflushed_chunk_timestamp_seconds",
+			Help: "Unix timestamp of the oldest unflushed chunk in the memory",
+		}),
 	}
 
-	if r != nil {
-		r.MustRegister(
-			m.flushQueueLength,
-			m.ingestedSamples,
-			m.ingestedSamplesFail,
-			m.queries,
-			m.queriedSamples,
-			m.queriedSeries,
-			m.queriedChunks,
-			m.memSeries,
-			m.memUsers,
-			m.walReplayDuration,
-			m.walCorruptionsTotal,
-		)
-
-		if registerMetricsConflictingWithTSDB {
-			r.MustRegister(
-				m.memSeriesCreatedTotal,
-				m.memSeriesRemovedTotal,
-			)
-		}
+	if registerMetricsConflictingWithTSDB {
```

Comment:
Previously these metrics were always created but not registered, just for safety. Now they are not created at all if `registerMetricsConflictingWithTSDB` is false. I've checked access to these two metrics, and this change looks safe to me... that is, these two metrics are not accessed, if metrics were created without them.

Should we rename `registerMetricsConflictingWithTSDB` now?

Reply:
Wow, very deep though. I agree renaming it. Done.